### ISSUE TITLE
Weird span positions for static recursion gate error

### ISSUE THREAD

**Issue by est31** [rust-lang/rust#39865]
If you consider the following example:

```rust
struct StaticDoubleLinked {
    prev: &'static StaticDoubleLinked,
    next: &'static StaticDoubleLinked,
    data: i32,
    head: bool,
}

static L1: StaticDoubleLinked = StaticDoubleLinked{prev: &L3, next: &L2, data: 1, head: true};
static L2: StaticDoubleLinked = StaticDoubleLinked{prev: &L1, next: &L3, data: 2, head: false};
static L3: StaticDoubleLinked = StaticDoubleLinked{prev: &L2, next: &L1, data: 3, head: false};

fn main() {}
```

Then it generates three gate errors. The problem is that all three errors point to the line where L1 is defined. This is bad. cc #39834

**Comment by est31** [comment]
cc also #29719

**Comment by nrc** [comment]
Closed by #40027
